File: python/snake_game/menu.py
"""wraps pygame_menu; settings live on GameSettings, not module globals like the old code"""
from dataclasses import dataclass

import pygame_menu

from . import config


@dataclass
class GameSettings:
    apple_count: int = config.FEW_APPLES
    ai_enabled: bool = False


class MenuController:
    def __init__(self, surface, settings: GameSettings):
        self.surface = surface
        self.settings = settings
        self._theme = pygame_menu.themes.Theme(
            title_font=pygame_menu.font.FONT_8BIT,
            background_color=(255, 0, 0, 0),
            widget_font=pygame_menu.font.FONT_8BIT,
        )
        self._start_menu = self._build_start_menu()

    def _build_start_menu(self):
        menu = pygame_menu.menu.Menu(height=300, width=400, theme=self._theme, title='Snake')
        menu.add.selector(
            'Apples: ', [('Little', config.FEW_APPLES), ('Many', config.MANY_APPLES)],
            onchange=lambda _, value: setattr(self.settings, 'apple_count', value),
        )
        menu.add.selector(
            'Player: ', [('Human', False), ('AI', True)],
            default=1 if self.settings.ai_enabled else 0,
            onchange=lambda _, value: setattr(self.settings, 'ai_enabled', value),
        )
        menu.add.button('Play', menu.disable)
        menu.add.button('Quit', pygame_menu.events.EXIT)
        return menu

    def show_start(self):
        self._start_menu.enable()
        self._start_menu.mainloop(self.surface)

    def show_game_over(self, score, high_score):
        menu = pygame_menu.menu.Menu(height=300, width=400, theme=self._theme, title='Game Over')
        menu.add.label(f'Score: {score}')
        menu.add.label(f'Best: {high_score}')
        menu.add.button('Play Again', menu.disable)
        menu.add.button('Quit', pygame_menu.events.EXIT)
        menu.mainloop(self.surface)

File: cpp/src/menu.cpp
#include "menu.hpp"

#include <raylib.h>

namespace {

constexpr Color kBackground{30, 30, 30, 255};
constexpr Color kNormal{255, 255, 255, 255};
constexpr Color kSelected{255, 215, 0, 255};
constexpr Color kHint{160, 160, 160, 255};

void DrawCentered(const char* text, int y, int fontSize, Color color) {
    int width = MeasureText(text, fontSize);
    DrawText(text, (config::WindowWidth - width) / 2, y, fontSize, color);
}

bool TogglePressed() {
    return IsKeyPressed(KEY_LEFT) || IsKeyPressed(KEY_RIGHT) || IsKeyPressed(KEY_A) || IsKeyPressed(KEY_D);
}

bool ConfirmPressed() {
    return IsKeyPressed(KEY_ENTER) || IsKeyPressed(KEY_SPACE) || IsKeyPressed(KEY_KP_ENTER);
}

}  // namespace

bool MenuController::ShowStart(GameSettings& settings) const {
    constexpr int kItemCount = 4;  // apples, player, play, quit
    int focus = 2;                 // start on "Play" so Enter alone is enough

    while (!WindowShouldClose()) {
        if (IsKeyPressed(KEY_DOWN) || IsKeyPressed(KEY_S)) focus = (focus + 1) % kItemCount;
        if (IsKeyPressed(KEY_UP) || IsKeyPressed(KEY_W)) focus = (focus - 1 + kItemCount) % kItemCount;

        if (focus == 0 && TogglePressed()) {
            settings.appleCount =
                (settings.appleCount == config::FewApples) ? config::ManyApples : config::FewApples;
        }
        if (focus == 1 && TogglePressed()) {
            settings.aiEnabled = !settings.aiEnabled;
        }
        if (ConfirmPressed()) {
            if (focus == 2) return true;   // Play
            if (focus == 3) return false;  // Quit
        }

        BeginDrawing();
        ClearBackground(kBackground);

        DrawCentered("SNAKE", 70, 56, kNormal);

        const char* appleLabel =
            settings.appleCount == config::FewApples ? "Apples:  <  Little  >" : "Apples:  <  Many  >";
        DrawCentered(appleLabel, 220, 28, focus == 0 ? kSelected : kNormal);

        const char* playerLabel = settings.aiEnabled ? "Player:  <  AI  >" : "Player:  <  Human  >";
        DrawCentered(playerLabel, 270, 28, focus == 1 ? kSelected : kNormal);

        DrawCentered("Play", 340, 32, focus == 2 ? kSelected : kNormal);
        DrawCentered("Quit", 390, 32, focus == 3 ? kSelected : kNormal);

        DrawCentered("Arrows to navigate, Enter to select", config::WindowHeight - 60, 18, kHint);

        EndDrawing();
    }
    return false;
}

bool MenuController::ShowGameOver(int score, int highScore) const {
    int focus = 0;  // 0 = play again, 1 = quit

    while (!WindowShouldClose()) {
        if (IsKeyPressed(KEY_DOWN) || IsKeyPressed(KEY_UP) || IsKeyPressed(KEY_S) || IsKeyPressed(KEY_W)) {
            focus = 1 - focus;
        }
        if (ConfirmPressed()) {
            return focus == 0;
        }

        BeginDrawing();
        ClearBackground(kBackground);

        DrawCentered("GAME OVER", 140, 48, kNormal);
        DrawCentered(TextFormat("Score: %d", score), 250, 28, kNormal);
        DrawCentered(TextFormat("Best: %d", highScore), 290, 28, kNormal);

        DrawCentered("Play Again", 380, 32, focus == 0 ? kSelected : kNormal);
        DrawCentered("Quit", 430, 32, focus == 1 ? kSelected : kNormal);

        DrawCentered("Arrows to navigate, Enter to select", config::WindowHeight - 60, 18, kHint);

        EndDrawing();
    }
    return false;
}
